File: src/services/index/context-pack.ts
import { invoke } from '@tauri-apps/api/core';

/** 上下文包条目 */
export interface ContextPackItem {
  id: string;
  // 来源文档（所有粒度都有）
  document_path: string;
  // 文档级字段
  title: string | null;
  summary: string | null;
  tags: string | null;
  // 细粒度可选字段（段落/选区时才有值）
  content: string | null;
  heading: string | null;
  start_line: number | null;
  end_line: number | null;
  // M4: chunk 关联字段
  chunk_id: number | null;
  source_type: string | null;
  score: number | null;
  reasoning_note: string | null;
  // 通用字段
  selected_reason: string;
  is_suggestion: boolean;
  timestamp: string;
}

/** 上下文包 */
export interface ContextPack {
  id: string;
  name: string;
  items: ContextPackItem[];
  created_at: string;
  updated_at: string;
}

export const contextPackService = {
  async listContextPacks(vaultPath: string): Promise<ContextPack[]> {
    return invoke('list_context_packs', { vaultPath });
  },

  async getContextPack(vaultPath: string, packId: string): Promise<ContextPack | null> {
    return invoke('get_context_pack', { vaultPath, packId });
  },

  async createContextPack(vaultPath: string, name: string): Promise<ContextPack> {
    return invoke('create_context_pack', { vaultPath, name });
  },

  async updateContextPack(
    vaultPath: string,
    packId: string,
    name?: string,
    items?: ContextPackItem[],
  ): Promise<ContextPack> {
    return invoke('update_context_pack', { vaultPath, packId, name: name ?? null, items: items ?? null });
  },

  async deleteContextPack(vaultPath: string, packId: string): Promise<void> {
    return invoke('delete_context_pack', { vaultPath, packId });
  },

  async exportContextPackMarkdown(vaultPath: string, packId: string): Promise<string> {
    return invoke('export_context_pack_markdown', { vaultPath, packId });
  },

  /** 前端辅助：添加条目到 pack */
  async addItem(
    vaultPath: string,
    packId: string,
    item: Omit<ContextPackItem, 'id' | 'timestamp'>,
  ): Promise<ContextPack> {
    const pack = await contextPackService.getContextPack(vaultPath, packId);
    if (!pack) throw new Error(`未找到 id 为 '${packId}' 的 context pack`);
    const newItem: ContextPackItem = {
      ...item,
      id: `${Date.now()}-${Math.random().toString(36).slice(2, 9)}`,
      timestamp: new Date().toISOString(),
    };
    pack.items.push(newItem);
    return contextPackService.updateContextPack(vaultPath, packId, undefined, pack.items);
  },

  /** 前端辅助：从 pack 移除条目 */
  async removeItem(vaultPath: string, packId: string, itemId: string): Promise<ContextPack> {
    const pack = await contextPackService.getContextPack(vaultPath, packId);
    if (!pack) throw new Error(`未找到 id 为 '${packId}' 的 context pack`);
    pack.items = pack.items.filter(i => i.id !== itemId);
    return contextPackService.updateContextPack(vaultPath, packId, undefined, pack.items);
  },

  /** 前端辅助：更新条目 */
  async updateItem(
    vaultPath: string,
    packId: string,
    itemId: string,
    updates: Partial<ContextPackItem>,
  ): Promise<ContextPack> {
    const pack = await contextPackService.getContextPack(vaultPath, packId);
    if (!pack) throw new Error(`未找到 id 为 '${packId}' 的 context pack`);
    const item = pack.items.find(i => i.id === itemId);
    if (!item) throw new Error(`未找到条目 '${itemId}'`);
    Object.assign(item, updates);
    return contextPackService.updateContextPack(vaultPath, packId, undefined, pack.items);
  },

  /** 前端辅助：重排条目 */
  async reorderItems(vaultPath: string, packId: string, itemIds: string[]): Promise<ContextPack> {
    const pack = await contextPackService.getContextPack(vaultPath, packId);
    if (!pack) throw new Error(`未找到 id 为 '${packId}' 的 context pack`);
    const reordered: ContextPackItem[] = [];
    for (const id of itemIds) {
      const item = pack.items.find(i => i.id === id);
      if (item) reordered.push(item);
    }
    pack.items = reordered;
    return contextPackService.updateContextPack(vaultPath, packId, undefined, pack.items);
  },

  /** 前端辅助：移动条目（上/下） */
  async moveItem(
    vaultPath: string,
    packId: string,
    itemId: string,
    direction: 'up' | 'down',
  ): Promise<ContextPack> {
    const pack = await contextPackService.getContextPack(vaultPath, packId);
    if (!pack) throw new Error(`未找到 id 为 '${packId}' 的 context pack`);
    const idx = pack.items.findIndex(i => i.id === itemId);
    if (idx === -1) return pack;
    const newIdx = direction === 'up' ? idx - 1 : idx + 1;
    if (newIdx < 0 || newIdx >= pack.items.length) return pack;
    const [moved] = pack.items.splice(idx, 1);
    pack.items.splice(newIdx, 0, moved);
    return contextPackService.updateContextPack(vaultPath, packId, undefined, pack.items);
  },

  /** 前端辅助：接受建议（将 is_suggestion 设为 false） */
  async acceptSuggestion(vaultPath: string, packId: string, itemId: string): Promise<ContextPack> {
    return contextPackService.updateItem(vaultPath, packId, itemId, { is_suggestion: false });
  },

  /** 前端辅助：重命名 pack */
  async renameContextPack(vaultPath: string, packId: string, name: string): Promise<ContextPack> {
    return contextPackService.updateContextPack(vaultPath, packId, name);
  },

  /** 前端辅助：导出为 Markdown（前端生成，含 source references） */
  exportAsMarkdown(pack: ContextPack): string {
    const lines: string[] = [];
    lines.push(`# ${pack.name}`);
    lines.push('');
    lines.push(`> 创建时间: ${new Date(pack.created_at).toLocaleString('zh-CN')}`);
    lines.push(`> 更新时间: ${new Date(pack.updated_at).toLocaleString('zh-CN')}`);
    lines.push(`> 内容数: ${pack.items.filter(i => !i.is_suggestion).length}`);
    lines.push('');

    const confirmedItems = pack.items.filter(i => !i.is_suggestion);
    for (const item of confirmedItems) {
      const displayTitle = item.heading ?? item.title ?? item.document_path.split('/').pop()?.replace('.md', '') ?? '未命名';
      lines.push(`## ${displayTitle}`);
      lines.push(`> 来源: ${item.document_path}`);
      if (item.start_line != null && item.end_line != null) {
        lines.push(`> 位置: L${item.start_line}–L${item.end_line}`);
      }
      if (item.summary) {
        lines.push(`> 摘要: ${item.summary}`);
      }
      if (item.tags) {
        lines.push(`> 标签: ${item.tags}`);
      }
      lines.push(`> 加入原因: ${item.selected_reason}`);
      lines.push('');
      if (item.content) {
        lines.push(item.content);
        lines.push('');
      }
      lines.push('---');
      lines.push('');
    }

    return lines.join('\n');
  },
};

File: src-tauri/src/commands/context_pack.rs
use crate::commands::vault::assert_path_inside_vault;
use chrono::Utc;
use serde::{Deserialize, Serialize};
use std::fs;
use std::path::Path;
use tauri::command;
use uuid::Uuid;

#[derive(Debug, Serialize, Deserialize, Clone)]
pub struct ContextPackItem {
    pub id: String,
    // 来源文档（所有粒度都有）
    pub document_path: String,
    // 文档级字段
    pub title: Option<String>,
    pub summary: Option<String>,
    pub tags: Option<String>,
    // 细粒度可选字段（段落/选区时才有值）
    pub content: Option<String>,
    pub heading: Option<String>,
    pub start_line: Option<i64>,
    pub end_line: Option<i64>,
    // M4: chunk 关联字段
    pub chunk_id: Option<i64>,
    pub source_type: Option<String>,
    pub score: Option<f64>,
    pub reasoning_note: Option<String>,
    // 通用字段
    pub selected_reason: String,
    pub is_suggestion: bool,
    pub timestamp: String,
}

#[derive(Debug, Serialize, Deserialize, Clone)]
pub struct ContextPack {
    pub id: String,
    pub name: String,
    pub items: Vec<ContextPackItem>,
    pub created_at: String,
    pub updated_at: String,
}

/// 获取 context-packs.json 文件路径
fn get_packs_path(vault: &Path) -> std::path::PathBuf {
    vault.join(".minddock").join("context-packs.json")
}

/// 读取所有 context packs
fn read_all_packs(vault: &Path) -> Result<Vec<ContextPack>, String> {
    let minddock_dir = vault.join(".minddock");
    let packs_path = get_packs_path(vault);

    // 路径校验
    let packs_path_str = packs_path.to_string_lossy().to_string();
    assert_path_inside_vault(
        &vault.to_string_lossy().to_string(),
        &packs_path_str,
    )?;

    fs::create_dir_all(&minddock_dir)
        .map_err(|e| format!("创建 .minddock 目录失败: {}", e))?;

    if !packs_path.exists() {
        return Ok(Vec::new());
    }

    let content = fs::read_to_string(&packs_path)
        .map_err(|e| format!("读取 context-packs.json 失败: {}", e))?;

    let packs: Vec<ContextPack> = serde_json::from_str(&content)
        .map_err(|e| format!("解析 context-packs.json 失败: {}", e))?;

    Ok(packs)
}

/// 安全写入所有 context packs（先写临时文件再 rename）
fn write_all_packs(vault: &Path, packs: &[ContextPack]) -> Result<(), String> {
    let minddock_dir = vault.join(".minddock");
    let packs_path = get_packs_path(vault);

    // 路径校验
    let packs_path_str = packs_path.to_string_lossy().to_string();
    assert_path_inside_vault(
        &vault.to_string_lossy().to_string(),
        &packs_path_str,
    )?;

    fs::create_dir_all(&minddock_dir)
        .map_err(|e| format!("创建 .minddock 目录失败: {}", e))?;

    let json = serde_json::to_string_pretty(packs)
        .map_err(|e| format!("序列化 context packs 失败: {}", e))?;

    let tmp_path = minddock_dir.join("context-packs.json.tmp");
    fs::write(&tmp_path, &json)
        .map_err(|e| format!("写入临时文件失败: {}", e))?;
    fs::rename(&tmp_path, &packs_path)
        .map_err(|e| format!("重命名临时文件失败: {}", e))?;

    Ok(())
}

#[command]
pub fn create_context_pack(
    vault_path: String,
    name: String,
) -> Result<ContextPack, String> {
    let vault = Path::new(&vault_path);
    let mut packs = read_all_packs(vault)?;

    let now = Utc::now().to_rfc3339();
    let pack = ContextPack {
        id: Uuid::new_v4().to_string(),
        name,
        items: Vec::new(),
        created_at: now.clone(),
        updated_at: now,
    };

    packs.push(pack.clone());
    write_all_packs(vault, &packs)?;

    Ok(pack)
}

#[command]
pub fn update_context_pack(
    vault_path: String,
    pack_id: String,
    name: Option<String>,
    items: Option<Vec<ContextPackItem>>,
) -> Result<ContextPack, String> {
    let vault = Path::new(&vault_path);
    let mut packs = read_all_packs(vault)?;

    let pack = packs
        .iter_mut()
        .find(|p| p.id == pack_id)
        .ok_or_else(|| format!("未找到 id 为 '{}' 的 context pack", pack_id))?;

    if let Some(n) = name {
        pack.name = n;
    }
    if let Some(i) = items {
        pack.items = i;
    }
    pack.updated_at = Utc::now().to_rfc3339();

    let updated = pack.clone();
    write_all_packs(vault, &packs)?;

    Ok(updated)
}

#[command]
pub fn delete_context_pack(
    vault_path: String,
    pack_id: String,
) -> Result<(), String> {
    let vault = Path::new(&vault_path);
    let mut packs = read_all_packs(vault)?;

    let original_len = packs.len();
    packs.retain(|p| p.id != pack_id);

    if packs.len() == original_len {
        return Err(format!("未找到 id 为 '{}' 的 context pack", pack_id));
    }

    write_all_packs(vault, &packs)?;

    Ok(())
}

#[command]
pub fn list_context_packs(
    vault_path: String,
) -> Result<Vec<ContextPack>, String> {
    let vault = Path::new(&vault_path);
    read_all_packs(vault)
}

#[command]
pub fn get_context_pack(
    vault_path: String,
    pack_id: String,
) -> Result<Option<ContextPack>, String> {
    let vault = Path::new(&vault_path);
    let packs = read_all_packs(vault)?;

    Ok(packs.into_iter().find(|p| p.id == pack_id))
}

#[command]
pub fn export_context_pack_markdown(
    vault_path: String,
    pack_id: String,
) -> Result<String, String> {
    let vault = Path::new(&vault_path);
    let packs = read_all_packs(vault)?;

    let pack = packs
        .into_iter()
        .find(|p| p.id == pack_id)
        .ok_or_else(|| format!("未找到 id 为 '{}' 的 context pack", pack_id))?;

    let mut md = String::new();
    md.push_str(&format!("# {}\n\n", pack.name));

    for item in &pack.items {
        md.push_str(&format!("> 来源: {}\n", item.document_path));
        if let Some(ref title) = item.title {
            md.push_str(&format!("> 标题: {}\n", title));
        }
        if let Some(ref heading) = item.heading {
            md.push_str(&format!("> 章节: {}\n", heading));
        }
        if let (Some(start), Some(end)) = (item.start_line, item.end_line) {
            md.push_str(&format!("> 位置: L{}–L{}\n", start, end));
        }
        if let Some(ref summary) = item.summary {
            md.push_str(&format!("> 摘要: {}\n", summary));
        }
        if let Some(ref tags) = item.tags {
            md.push_str(&format!("> 标签: {}\n", tags));
        }
        md.push_str(&format!("> 加入时间: {}\n", item.timestamp));
        md.push_str(&format!("> 加入原因: {}\n\n", item.selected_reason));
        if let Some(ref content) = item.content {
            md.push_str(content);
            md.push_str("\n---\n\n");
        } else {
            md.push_str("---\n\n");
        }
    }

    Ok(md)
}
